### PR TITLE
Version 4.0.0 - API redesign - no TH + native ALB support

### PR DESCRIPTION
**Pull request by dnikolovv**
Since https://github.com/theam/aws-lambda-haskell-runtime/pull/81 went stale and this is a needed feature, after a discussion with @NickSeagull we decided that I could tackle it.

This is version `4.0.0` - the API is completely redesigned and you no longer need template Haskell. Also, you can write your handler code in any monad.

You can now do this:

```haskell
initializeConfig :: IO Config
initializeConfig = implementation

main :: IO ()
main = do
  let options = defaultDispatcherOptions

  -- "id" is being used as `m a -> IO a`
  runLambdaHaskellRuntime options initializeConfig id $ do
    addALBHandler "handler" $ \(request :: ALBRequest someRequestType) (context :: IORef Config) ->

  -- or if you want to use API Gateway
  runLambdaHaskellRuntime options initializeConfig id $ do
    addAPIGatewayHandler "handler" $ \(request :: ApiGateway someRequestType) (context :: IORef Config) ->

  -- or if you want to use a standalone lambda
  runLambdaHaskellRuntime options initializeConfig id $ do
    addStandaloneLambdaHandler "handler" $ \(request :: someRequestType) (context :: IORef Config) ->
```

Using [aws-lambda-haskell-runtime-wai](https://github.com/eir-forsakring/aws-lambda-haskell-runtime-wai), you can now support both API Gateway and ALB like so:

```haskell
mkApp :: IO Wai.Application
mkApp = implementation

main :: IO ()
main = do
  let options = defaultDispatcherOptions

  runWaiAsLambda
    options
    "handler"
    -- If you want to load-balance multiple lambdas using the same balancer, for each lambda you can
    -- specify the ignored path part so your routes don't break.
    (ignoreALBPathPart "my-lambda")
    mkApp
```

- [x] - Bump version to `4.0.0`, update to latest stable resolver
- [x] - Add `-Werror` and `-Wall`, clean up
- [x] - Use ormolu for consistent formatting
- [x] - Remove TH code
- [x] - Develop a new "run" API
- [x] - Enable users to write their handlers in any monad
- [x] - Update [aws-lambda-haskell-runtime-wai](https://github.com/eir-forsakring/aws-lambda-haskell-runtime-wai)
- [x] - Document

### CODE REVIEW

**Review by NickSeagull — COMMENTED** [review]
Thanks! Left some comments, will give a more thorough look next year (lol)

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "haskell.serverExecutablePath": "/home/random/.local/bin/hls-8.8.4/haskell-language-server-8.8.4"
```
[the remark this answers is not in the document]
Comment (dnikolovv) [reply .vscode/settings.json:2]: Yeah I'll ignore it.

File: aws-lambda-haskell-runtime.cabal [hunk 2]
```diff
@@ -1,19 +1,21 @@
 cabal-version: 1.12
 
--- This file has been generated from package.yaml by hpack version 0.34.2.
+-- This file has been generated from package.yaml by hpack version 0.33.0.
```
[the remark this answers is not in the document]
Comment (dnikolovv) [reply aws-lambda-haskell-runtime.cabal:3]: Makes sense, I'll switch to Cabal.

File: src/Aws/Lambda/Runtime.hs [hunk 3]
```diff
@@ -29,84 +18,98 @@ import qualified Aws.Lambda.Runtime.Environment as Environment
...
+import Data.Aeson (ToJSON (toJSON))
+import Data.IORef (newIORef)
+import Data.Text (Text, unpack)
+import qualified Network.HTTP.Client as Http
+import System.IO (hFlush, stderr, stdout)
 
 -- | Runs the user @haskell_lambda@ executable and posts back the
 -- results. This is called from the layer's @main@ function.
-runLambda :: forall context. IO context -> Runtime.RunCallback context -> IO ()
+runLambda :: forall context t. IO context -> Runtime.RunCallback t context -> IO ()
```
Comment (NickSeagull) [inline src/Aws/Lambda/Runtime.hs:32]: What is `t` in this case? Could you give it a more semantic name please?

File: src/Aws/Lambda/Runtime.hs [hunk 4]
```diff
@@ -29,84 +18,98 @@ import qualified Aws.Lambda.Runtime.Environment as Environment
...
-  :: Throws Error.Invocation
-  => Throws Error.EnvironmentVariableNotSet
-  => Runtime.RunCallback context
-  -> ApiInfo.Event
-  -> Context.Context context
-  -> IO Runtime.LambdaResult
+invokeWithCallback ::
+  Throws Error.Invocation =>
+  Throws Error.EnvironmentVariableNotSet =>
+  Runtime.RunCallback t context ->
```
Comment (NickSeagull) [inline src/Aws/Lambda/Runtime.hs:79]: Can you give `t` a more semantic name please?

File: src/Aws/Lambda/Runtime/API/Endpoints.hs [hunk 5]
```diff
@@ -1,61 +1,67 @@
...
 nextInvocation lambdaApi =
-  Endpoint $ concat
-    [ "http://"
-    , lambdaApi
-    , "/"
-    , Version.value
-    , "/runtime/invocation/next"
-    ]
+  Endpoint $
+    mconcat
```
Comment (NickSeagull) [inline src/Aws/Lambda/Runtime/API/Endpoints.hs:21]: Lovely change to be more abstract ❤️

File: src/Aws/Lambda/Runtime/Common.hs [hunk 6]
```diff
@@ -1,90 +1,73 @@
...
-  UseWithAPIGateway |
-  StandaloneLambda
-  deriving (Lift)
+import Data.String (IsString)
 
 -- | Callback that we pass to the dispatcher function
-type RunCallback context =
-  LambdaOptions context -> IO (Either LambdaError LambdaResult)
+type RunCallback (t :: HandlerType) context =
+  LambdaOptions context -> IO (Either (LambdaError t) (LambdaResult t))
```
Comment (NickSeagull) [inline src/Aws/Lambda/Runtime/Common.hs:39]: Oh, I see now where the `t` comes from 😅 

Still, even though it seems redundant, I'd call this like

```suggestion
type RunCallback (handlerType :: HandlerType) context =
  LambdaOptions context -> IO (Either (LambdaError handlerType) (LambdaResult handlerType))
```

File: src/Aws/Lambda/Setup.hs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+  )
+import Data.Aeson (FromJSON)
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Text as Text
+import Data.Typeable (Typeable)
+import GHC.IO.Handle.FD (stderr)
+import GHC.IO.Handle.Text (hPutStr)
+
+type Handlers t m context request response error =
+  HM.HashMap HandlerName (Handler t m context request response error)
```
Comment (NickSeagull) [inline src/Aws/Lambda/Setup.hs:69]: Same with `t` here :)

File: src/Aws/Lambda/Setup.hs [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+type ALBCallback m context request response error =
+  (ALBRequest request -> Context context -> m (Either (ALBResponse error) (ALBResponse response)))
+
+data Handler (t :: HandlerType) m context request response error where
+  StandaloneLambdaHandler :: StandaloneCallback m context request response error -> Handler 'StandaloneHandlerType m context request response error
+  APIGatewayHandler :: APIGatewayCallback m context request response error -> Handler 'APIGatewayHandlerType m context request response error
+  ALBHandler :: ALBCallback m context request response error -> Handler 'ALBHandlerType m context request response error
+
+newtype HandlersM (t :: HandlerType) m context request response error a = HandlersM
```
Comment (NickSeagull) [inline src/Aws/Lambda/Setup.hs:85]: `t` here

File: src/Aws/Lambda/Setup.hs [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+  {runHandlersM :: StateT (Handlers t m context request response error) IO a}
+  deriving newtype
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadState (Handlers t m context request response error)
+    )
+
+type RuntimeContext (t :: HandlerType) m context request response error =
```
Comment (NickSeagull) [inline src/Aws/Lambda/Setup.hs:94]: `t` here

File: src/Aws/Lambda/Setup.hs [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    ToALBResponseBody error,
+    ToALBResponseBody response,
+    FromJSON (ApiGatewayRequest request),
+    FromJSON (ALBRequest request),
+    FromJSON request,
+    Typeable request
+  )
+
+runLambdaHaskellRuntime ::
+  RuntimeContext t m context request response error =>
```
Comment (NickSeagull) [inline src/Aws/Lambda/Setup.hs:110]: `t` here

File: src/Aws/Lambda/Setup.hs [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+addALBHandler ::
+  HandlerName ->
+  ALBCallback m context request response error ->
+  HandlersM 'ALBHandlerType m context request response error ()
+addALBHandler handlerName handler =
+  State.modify (HM.insert handlerName (ALBHandler handler))
+
+handlerToCallback ::
+  forall t m context request response error.
```
Comment (NickSeagull) [inline src/Aws/Lambda/Setup.hs:158]: `t` here

**Review by NickSeagull — CHANGES_REQUESTED** [review]
Looks great! I just left some suggestions, let's see if we can get this merged soon.

Thanks a lot 🙏

File: src/Aws/Lambda/Runtime.hs (hunk 3 above)
Comment (dnikolovv) [reply src/Aws/Lambda/Runtime.hs:32]: Consider the `t` eliminated. :neckbeard:

File: aws-lambda-haskell-runtime.cabal (hunk 2 above)
[the remark this answers is not in the document]
Comment (dnikolovv) [reply aws-lambda-haskell-runtime.cabal]: Actually, after reading the issue - Stack isn't deprecating `package.yaml`, it's still quite useful as you don't need to specify each module name. You just need to include the generated `.cabal` file in the repo. I'll leave it and not ignore the `.cabal` anymore.

**Comment by NickSeagull** [comment]
Hey, what's the status on this?

**Comment by freuk** [comment]
This change looks great, I'm using it and am looking forwards to having it in the upcoming release if it gets merged. Thanks.

**Comment by dnikolovv** [comment]
The initial template has been updated and I'm halfway through the docs. Should finish it this Saturday.

**Comment by dnikolovv** [comment]
@NickSeagull The docs are really not perfect (or well done), but I guess we can roll with it and update those at a later stage.

**Comment by NickSeagull** [comment]
Thanks a lot!!!! ❤️

**Pull request merged**